Mathlib/RingTheory/AdicCompletion/Algebra.lean

```diff
@@ -0,0 +1,219 @@
+definitional behaviour for the module instance on adic completions -/
+instance : SMul (R ⧸ (I • ⊤ : Ideal R)) (M ⧸ (I • ⊤ : Submodule R M)) where
+  smul r x :=
+    Quotient.liftOn r (· • x) fun b₁ b₂ (h : Setoid.Rel _ b₁ b₂) ↦ by
+      refine Quotient.inductionOn' x (fun x ↦ ?_)
+      have h : b₁ - b₂ ∈ (I : Submodule R R) := by
+        rwa [show I = I • ⊤ by simp, ← Submodule.quotientRel_r_def]
+      rw [← sub_eq_zero, ← sub_smul, Submodule.Quotient.mk''_eq_mk,
+        ← Submodule.Quotient.mk_smul, Submodule.Quotient.mk_eq_zero]
+      exact Submodule.smul_mem_smul h mem_top
+
+@[local simp]
+theorem mk_smul_mk (r : R) (x : M) :
+    Ideal.Quotient.mk (I • ⊤) r • Submodule.Quotient.mk (p := (I • ⊤ : Submodule R M)) x
+      = r • Submodule.Quotient.mk (p := (I • ⊤ : Submodule R M)) x :=
+  rfl
+
+instance : Module (R ⧸ (I • ⊤ : Ideal R)) (M ⧸ (I • ⊤ : Submodule R M)) :=
+  Function.Surjective.moduleLeft (Ideal.Quotient.mk (I • ⊤ : Ideal R))
+    Ideal.Quotient.mk_surjective (fun _ _ ↦ rfl)
+
+instance : IsScalarTower R (R ⧸ (I • ⊤ : Ideal R)) (M ⧸ (I • ⊤ : Submodule R M)) where
+  smul_assoc r s x := by
+    refine Quotient.inductionOn' s (fun s ↦ ?_)
+    refine Quotient.inductionOn' x (fun x ↦ ?_)
+    simp only [Submodule.Quotient.mk''_eq_mk]
+    rw [← Submodule.Quotient.mk_smul, Ideal.Quotient.mk_eq_mk, mk_smul_mk, smul_assoc]
+    rfl
+
+instance smul : SMul (AdicCompletion I R) (AdicCompletion I M) where
+  smul r x := {
+    val := fun n ↦ eval I R n r • eval I M n x
+    property := fun {m n} hmn ↦ by
+      apply induction_on I R r (fun r ↦ ?_)
+      apply induction_on I M x (fun x ↦ ?_)
+      simp only [coe_eval, mk_apply_coe, mkQ_apply, Ideal.Quotient.mk_eq_mk,
+        mk_smul_mk, LinearMapClass.map_smul, transitionMap_mk]
+      rw [smul_mk I hmn]
+  }
+
+@[simp]
+theorem smul_eval (n : ℕ) (r : AdicCompletion I R) (x : AdicCompletion I M) :
+    (r • x).val n = r.val n • x.val n :=
+  rfl
+
+/-- `AdicCompletion I M` is naturally an `AdicCompletion I R` module. -/
+instance module : Module (AdicCompletion I R) (AdicCompletion I M) where
+  one_smul b := by
+    ext n
+    simp only [smul_eval, val_one, one_smul]
+  mul_smul r s x := by
+    ext n
+    simp only [smul_eval, val_mul, mul_smul]
+  smul_zero r := by
+    ext n
+    rw [smul_eval, val_zero, smul_zero]
+  smul_add r x y := by
+    ext n
+    simp only [smul_eval, val_add, smul_add]
+  add_smul r s x := by
+    ext n
+    simp only [coe_eval, smul_eval, map_add, add_smul, val_add]
+  zero_smul x := by
+    ext n
+    simp only [smul_eval, _root_.map_zero, zero_smul, val_zero]
+
+instance : IsScalarTower R (AdicCompletion I R) (AdicCompletion I M) where
+  smul_assoc r s x := by
+    ext n
+    rw [smul_eval, val_smul, val_smul, smul_eval, smul_assoc]
+
+/-- A priori `AdicCompletion I R` has two `AdicCompletion I R`-module instances.
+Both agree definitionally. -/
+theorem smul_eq_mul (r s : AdicCompletion I R) : r • s = r * s :=
```

Comment:
We usually write this kind of test as examples (we don't want them in the docs). Something like
```lean
example : module I = @Algebra.toModule (AdicCompletion I R)
  (AdicCompletion I R) _ _ (Algebra.id _) := rfl
```
Two comments:
* are these the two instantes you want to compare? It's better to write them like this than using `•` (we check all the structures at once).
* `rfl` works, but the real test is `by with_reducible rfl` which *does not* work (the difference is that `by with_reducible rfl` does not unfold definitions that are not marked `reducible)`. I am not 100% sure to understand all the subtleties, but this probably means that for Lean those are not exactly the same. @eric-wieser  will this be a problem?

Reply:
@eric-wieser do you think this is OK?

Reply:
I don't know if it helps, but `by with_reducible_and_instances rfl` works. I just updated the PR.

Reply:
Ah, I see, that's OK I think.